pypesto/ensemble/ensemble.py

```diff
@@ -367,6 +371,164 @@ def from_sample(
+        elif len(x_vectors) < max_size:
+            logger.info(f'The ensemble contains {len(x_vectors)} parameter '
+                        f'vectors, which is less than the maximum size. If '
+                        f'you want to include more vectors you can consider '
+                        f'raising the cutoff value or including parameters '
+                        f'from the history with the from_history() function.')
+
+        x_vectors = np.stack(x_vectors, axis=1)
+        return Ensemble(x_vectors=x_vectors,
+                        x_names=x_names,
+                        vector_tags=vector_tags,
+                        lower_bound=lb,
+                        upper_bound=ub,
+                        **kwargs)
+
+    @staticmethod
+    def from_history(
+            result: Result,
+            cutoff: float,
+            max_size: int,
+            max_per_start: int = None,
+            **kwargs,
+    ):
+        """Construct an ensemble from the history of an optimization.
+
+        Parameters
+        ----------
+        result:
+            A pyPESTO result that contains a sample result.
+        cutoff:
+            Exclude parameters from the optimization if the nllh
+            is higher than the `cutoff`.
+        max_size:
+            The maximum size the ensemble should be.
+        max_per_start:
+            The maxmimum number of vectors to be included from a
+            single optimization start.
+
+        Returns
+        -------
+        The ensemble.
+        """
+        if not result.optimize_result.list[0].history.options['trace_record']:
+            logger.warning('The optimize result has no trace. The Ensemble '
+                           'will automatically be created through '
+                           'from_optimization().')
+            return Ensemble.from_optimization(result=result,
+                                              cutoff=cutoff,
+                                              max_size=max_size, **kwargs)
+        x_vectors = []
+        vector_tags = []
+        x_names = result.problem.x_names
+        lb = result.problem.lb_full
+        ub = result.problem.ub_full
+
+        # calculate the number of starts whose final nllh is below cutoff
+        n_starts = 0
+        for start in result.optimize_result.list:
+            if start['fval'] < cutoff:
+                n_starts += 1
+            else:
+                break
+
+        if n_starts*max_per_start > max_size:
+            logger.info(f'The number of starts that can contribute an '
+                        f'ensemble vector muliplied with max_per_start '
+                        f'is higher than max_size. Thus we will lower '
+                        f'max_per_start. If you do not want this to '
+                        f'happen consider increasing max_size to '
+                        f'{max_per_start*n_starts} or decrease cutoff.')
+            max_per_start = math.floor(max_size/n_starts)
```

Comment:
Perhaps all possible vectors from the different starts should be collected first, then trimmed back until `max_size` is reached? e.g. given two history traces `trace_0` and `trace_1`, `max_size == 10` and `max_per_start == 10`, and given `len(sum(trace_0['fval'] <= cutoff)) == 7` and `len(sum(trace_1['fval'] <= cutoff)) == 3`, then I think the current implementation would result in an ensemble with 8 vectors, not 10? (ignoring that the current implementation stops at the first `fval` that is worse than `cutoff`)

Reply:
Yes, but in this scenario you would have 5 from `trace_0` and 3 from `trace_1` which is already leaning towards `trace_0`. This would be even stronger if we kept `max_per_start==10` in that case it would be `7:3`. Will put it as a discussion point for next meeting :)

Reply:
Ah, if I specify a `cutoff` and `max_per_start` then I would expect `7:3`. Any bias that arises from the users choices should probably be handled by the user too right? Otherwise the user may wonder why the ensemble contains fewer vectors than expected.

e.g. could add an option like `max_diff_between_starts = 2` to ensure the number of vectors from any start is within two of the number of vectors from the start that contributes the fewest vectors (then `7:3` will be truncated to `5:3`)

Reply:
I agree that just going for floor here is not ideal... This will result in (maybe much) smaller ensembles than necessary in some cases, which we don't want to happen. Another thing: Why going for `math.floor` rather than `np.floor`? Numpy is imported anyways, no need to import math as well...